File: 503-next-greater-element-ii/503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n=nums.length;

        int[] ngr=new int[n];
        Stack<Integer> st=new Stack<>();

        for(int i=n-1; i>=0; i--){
            int ele=nums[i];

            while(st.size()>0 && nums[st.peek()]<=ele){
                st.pop();
            }

            if(st.size()==0){
                ngr[i]=-1;
            } else {
                ngr[i]=nums[st.peek()];
            }

            st.push(i);
        }

        for(int i=n-1; i>=0; i--){
            int ele=nums[i];

            while(st.size()>0 && nums[st.peek()]<=ele){
                st.pop();
            }

            if(st.size()==0){
                ngr[i]=-1;
            } else {
                ngr[i]=nums[st.peek()];
            }

            st.push(i);
        }

        return ngr;
    }
}

File: 503-next-greater-element-ii/503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        stack<int> st;  
        int n=nums.size();

        vector<int> ans(n,-1);

        for(int i=0; i<2*n; i++){
            while(st.size()!=0 && nums[st.top()]<nums[i%n]){
                ans[st.top()]=nums[i%n];
                st.pop();
            }

            if(i<n)
               st.push(i);
        }      

        return ans;  
    }
};